### PR TITLE
Chantier IC + modernisation

### PR DESCRIPTION
**Pull request by rducom**
WIP

- restructuration solution (dossier src / tests)
- upgrade net461 
- upgrade xunit 2.3.1 
- upgrade RabbitMQ.Client 5.0.1
- migrate on Netstandard2.0
- plug new IC stack : AppVeyor + Codecov + sonarCloud
- automatic Version increment (more on https://github.com/rducom/ALM)

### CODE REVIEW

**Comment by BriceFrancois** [comment]
:x: [Build MerQure 1.0.75 failed](https://ci.appveyor.com/project/BriceFrancois/merqure/build/1.0.75) (commit https://github.com/LuccaSA/MerQure/commit/39db71bc51 by @rducom)

**Comment by bebelz** [comment]
appVeyor is unhappy :(

**Comment by BriceFrancois** [comment]
:x: [Build MerQure 1.0.76 failed](https://ci.appveyor.com/project/BriceFrancois/merqure/build/1.0.76) (commit https://github.com/LuccaSA/MerQure/commit/c6e7f5598c by @rducom)

File: src/MerQure.RbMQ/MessagingService.cs [hunk 1]
```diff
@@ -1,43 +1,58 @@
...
+    /// <summary>
+    /// Must be registered as a singleton
+    /// </summary>
+    public class SharedConnection
     {
-        private static IConnection GetRabbitMqConnection()
+        private readonly RabbitMqConnectionConfigurationElement _config;
+        private readonly Lazy<IConnection> _currentConnection;
+
+        public SharedConnection(IOptions<RabbitMqConnectionConfigurationElement> options)
```
Comment (BriceFrancois) [inline src/MerQure.RbMQ/MessagingService.cs]: IOptions est équivalent à Lazy ? 
"connectionConfig" ou "config" plutot que "options".

File: src/MerQure.RbMQ/MessagingService.cs [hunk 2]
```diff
@@ -1,43 +1,58 @@
...
         {
-            var rabbitMqConnection = Config.RabbitMqConfiguration.GetConfig().Connection;
+            _config = options.Value;
+            _currentConnection = new Lazy<IConnection>(() => GetRabbitMqConnection(_config));
+        }
 
+        public IConnection CurrentConnection => _currentConnection.Value;
```
Comment (BriceFrancois) [inline src/MerQure.RbMQ/MessagingService.cs]: Je ne vois pas comment on garantit le singleton ? On va créer une nouvelle instance à chaque fois.

File: src/MerQure.RbMQ/MessagingService.cs [hunk 3]
```diff
@@ -1,43 +1,58 @@
...
-        private static readonly Lazy<IConnection> currentConnection = new Lazy<IConnection>(GetRabbitMqConnection);
-        public static IConnection CurrentConnection => currentConnection.Value;
+    }
+
+    public class MessagingService : IMessagingService
+    {
+        public bool Durable => Config.Durable;
+        public bool AutoDeleteQueue => Config.AutoDeleteQueue;
+        public string ExchangeType { get; set; } = RabbitMQ.Client.ExchangeType.Topic;
+        public ushort DefaultPrefetchCount => Config.DefaultPrefetchCount;
```
Comment (BriceFrancois) [inline src/MerQure.RbMQ/MessagingService.cs]: Attention, le fait de charger la valeur depuis la config ne veit pas dire qu'on ne va pas la changer dans le service.

File: src/MerQure.RbMQ/MessagingService.cs [hunk 4]
```diff
@@ -1,43 +1,58 @@
...
-            var rabbitMqConfig = Config.RabbitMqConfiguration.GetConfig();
+        private readonly SharedConnection _sharedConnection;
+        private readonly IOptions<RabbitMqConfigurationSection> _rabbitMqConfig;
 
-            this.Durable = rabbitMqConfig.Durable;
-            this.AutoDeleteQueue = rabbitMqConfig.AutoDeleteQueue;
-            this.DefaultPrefetchCount = rabbitMqConfig.DefaultPrefetchCount;
-            this.PublisherAcknowledgementsTimeoutInMilliseconds = rabbitMqConfig.PublisherAcknowledgementsTimeoutInMilliseconds;
-            this.ExchangeType = RabbitMQ.Client.ExchangeType.Topic;
+        public MessagingService(IOptions<RabbitMqConfigurationSection> rabbitMqConfig, SharedConnection sharedConnection)
```
Comment (BriceFrancois) [inline src/MerQure.RbMQ/MessagingService.cs]: j'imagine que c'est la sharedConnection qui garantit le singleton,
du coup, la gestion Lazy dans la sharedConnection ne sert plus à rien non ?

File: tests/MerQure.RbMQ.Tests/MessagingServiceTests.cs [hunk 5]
```diff
@@ -8,7 +9,7 @@ public class MessagingServiceTests
         [Fact]
         public void DeclareExchangeFailWithEmptyName()
         {
-            var messagingService = new MessagingService();
+            var messagingService = new MessagingService(null,null);
 
             Assert.Throws<ArgumentNullException>(() => messagingService.DeclareExchange(null));
```
Comment (BriceFrancois) [inline tests/MerQure.RbMQ.Tests/MessagingServiceTests.cs]: on pourrait déclarer des configExceptions

File: src/MerQure.RbMQ/Config/RabbitMqConfigurationSection.cs [hunk 6]
```diff
@@ -1,13 +1,8 @@
...
-using System.Collections.Generic;
-using System.Configuration;
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
 
 namespace MerQure.RbMQ.Config
 {
-    public class RabbitMqConfigurationSection : ConfigurationSection
+    public class RabbitMqConfigurationSection
```
Comment (BriceFrancois) [inline src/MerQure.RbMQ/Config/RabbitMqConfigurationSection.cs]: nommage à réfléchir, puisqu'on utilise plus les configsections

File: src/MerQure.RbMQ/Config/RabbitMqConnectionConfigurationElement.cs [hunk 7]
```diff
@@ -1,13 +1,8 @@
...
-using System.Text;
-using System.Threading.Tasks;
 
 namespace MerQure.RbMQ.Config
 {
-    public class RabbitMqConnectionConfigurationElement : ConfigurationElement
+    public class RabbitMqConnectionConfigurationElement  
     {
         public const string DefaultConnectionString = "amqp://guest:guest@localhost:5672/";
         private const string ConnectionStringPropertyName = "connectionString";
```
Comment (LuccaIntegration) [inline src/MerQure.RbMQ/Config/RabbitMqConnectionConfigurationElement.cs]: MAJOR Remove the unused private field 'ConnectionStringPropertyName'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)

File: src/MerQure.RbMQ/Config/RabbitMqConfigurationSection.cs [hunk 8]
```diff
@@ -1,13 +1,8 @@
...
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
 
 namespace MerQure.RbMQ.Config
 {
-    public class RabbitMqConfigurationSection : ConfigurationSection
+    public class RabbitMqConfigurationSection 
     {
         private const string DurablePropertyName = "durable";
```
Comment (LuccaIntegration) [inline src/MerQure.RbMQ/Config/RabbitMqConfigurationSection.cs]: MAJOR Remove the unused private field 'DurablePropertyName'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)

File: src/MerQure.RbMQ/Config/RabbitMqConfigurationSection.cs [hunk 9]
```diff
@@ -1,13 +1,8 @@
...
         private const string AutoDeleteQueuePropertyName = "autoDeleteQueue";
```
Comment (LuccaIntegration) [inline src/MerQure.RbMQ/Config/RabbitMqConfigurationSection.cs]: MAJOR Remove the unused private field 'AutoDeleteQueuePropertyName'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)

**Comment by LuccaIntegration** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 7 major

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [RabbitMqConfigurationSection.cs#L9](https://github.com/LuccaSA/MerQure/blob/acb34bf0b07dbd7d71a9479d75a3e10fa55832d8/src/MerQure.RbMQ/Config/RabbitMqConfigurationSection.cs#L9): Remove the unused private field 'DefaultPrefetchCountPropertyName'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)
1. ![MAJOR][MAJOR] [RabbitMqConfigurationSection.cs#L10](https://github.com/LuccaSA/MerQure/blob/acb34bf0b07dbd7d71a9479d75a3e10fa55832d8/src/MerQure.RbMQ/Config/RabbitMqConfigurationSection.cs#L10): Remove the unused private field 'PublisherAcknowledgementsTimeoutPropertyName'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)
1. ![MAJOR][MAJOR] [RabbitMqConnectionConfigurationElement.cs#L9](https://github.com/LuccaSA/MerQure/blob/acb34bf0b07dbd7d71a9479d75a3e10fa55832d8/src/MerQure.RbMQ/Config/RabbitMqConnectionConfigurationElement.cs#L9): Remove the unused private field 'AutomaticRecoveryEnabledPropertyName'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)
1. ![MAJOR][MAJOR] [RabbitMqConnectionConfigurationElement.cs#L10](https://github.com/LuccaSA/MerQure/blob/acb34bf0b07dbd7d71a9479d75a3e10fa55832d8/src/MerQure.RbMQ/Config/RabbitMqConnectionConfigurationElement.cs#L10): Remove the unused private field 'TopologyRecoveryEnabledPropertyName'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/MerQure.RbMQ/MessagingService.cs (hunk 4 above)
Comment (rducom) [reply src/MerQure.RbMQ/MessagingService.cs]: Soit je créer la connection dans le contructeur, et on aura une éventuelle exception lors de l'injection (ce qui est pas top), soit je créer a connection lors de l'appel, et dans le cas on peut avoir un problème de concurrence (plusieurs appels de création de la connection en //). Le lazy dans ce cas permet de garantir une seule execution.

File: src/MerQure.RbMQ/Config/MerQureConnection.cs [hunk 10]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+
+namespace MerQure.RbMQ.Config
+{
+    public class MerQureConnection
+    {
+        public const string DefaultConnectionString = "amqp://guest:guest@localhost:5672/";
+        private const string AutomaticRecoveryEnabledPropertyName = "automaticRecoveryEnabled";
```
Comment (LuccaIntegration) [inline src/MerQure.RbMQ/Config/MerQureConnection.cs]: MAJOR Remove the unused private field 'AutomaticRecoveryEnabledPropertyName'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)

File: src/MerQure.RbMQ/Config/MerQureConnection.cs [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+        private const string TopologyRecoveryEnabledPropertyName = "topologyRecoveryEnabled";
```
Comment (LuccaIntegration) [inline src/MerQure.RbMQ/Config/MerQureConnection.cs]: MAJOR Remove the unused private field 'TopologyRecoveryEnabledPropertyName'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)

File: src/MerQure.RbMQ/Config/MerQureConfiguration.cs [hunk 12]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+
+namespace MerQure.RbMQ.Config
+{
+    public class MerQureConfiguration
+    {
+        private const string DefaultPrefetchCountPropertyName = "defaultPrefetchCount";
```
Comment (LuccaIntegration) [inline src/MerQure.RbMQ/Config/MerQureConfiguration.cs]: MAJOR Remove the unused private field 'DefaultPrefetchCountPropertyName'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)

File: src/MerQure.RbMQ/Config/MerQureConfiguration.cs [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+        private const string PublisherAcknowledgementsTimeoutPropertyName = "publisherAcknowledgementsTimeoutInMilliseconds";
```
Comment (LuccaIntegration) [inline src/MerQure.RbMQ/Config/MerQureConfiguration.cs]: MAJOR Remove the unused private field 'PublisherAcknowledgementsTimeoutPropertyName'. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1144)

File: src/MerQure.RbMQ/MessagingService.cs [hunk 14]
```diff
@@ -7,37 +9,26 @@ namespace MerQure.RbMQ
...
+        public bool Durable { get; set; }
+        public bool AutoDeleteQueue { get; set; }
         public ushort DefaultPrefetchCount { get; set; }
         public long PublisherAcknowledgementsTimeoutInMilliseconds { get; set; }
 
-        public MessagingService()
+        protected IConnection CurrentConnection => _sharedConnection.CurrentConnection;
+
+        private readonly SharedConnection _sharedConnection;
+        private readonly IOptions<MerQureConfiguration> _merQureConfiguration;
```
Comment (LuccaIntegration) [inline src/MerQure.RbMQ/MessagingService.cs]: MINOR Remove the '_merQureConfiguration' field and declare it as a local variable in the relevant methods. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1450)

**Comment by LuccaIntegration** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/MerQure.RbMQ/MessagingService.cs [hunk 15]
```diff
@@ -7,37 +9,30 @@ namespace MerQure.RbMQ
...
-        }
-        private static readonly Lazy<IConnection> currentConnection = new Lazy<IConnection>(GetRabbitMqConnection);
-        public static IConnection CurrentConnection => currentConnection.Value;
-
-        public bool Durable { get; private set; }
-        public bool AutoDeleteQueue { get; private set; }
-        public string ExchangeType { get; private set; }
+        public string ExchangeType { get; set; } = RabbitMQ.Client.ExchangeType.Topic;
+        public bool Durable { get; set; }
+        public bool AutoDeleteQueue { get; set; }
```
Comment (BriceFrancois) [inline src/MerQure.RbMQ/MessagingService.cs]: je disais de les laisser modifiables, mais juste ceux qui l'étaient déjà :( désolé

File: src/MerQure.RbMQ/Config/MerQureConfiguration.cs [hunk 16]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+
+namespace MerQure.RbMQ.Config
+{
+    public class MerQureConfiguration
```
Comment (BriceFrancois) [inline src/MerQure.RbMQ/Config/MerQureConfiguration.cs]: j'aurai laissé RabbitMqConfig, mais ça me va comme ça.

**Pull request merged**